test/unit/org/apache/cassandra/index/sai/functional/GroupComponentsTest.java

```diff
@@ -0,0 +1,113 @@
+
+public class GroupComponentsTest extends SAITester
+{
+    @Test
+    public void testInvalidateWithoutObsolete() throws Throwable
+    {
+        createTable("CREATE TABLE %s (pk int primary key, value text)");
+        createIndex("CREATE CUSTOM INDEX ON %s(value) USING 'StorageAttachedIndex'");
+        waitForIndexQueryable();
+        execute("INSERT INTO %s (pk) VALUES (1)");
+        flush();
+
+        ColumnFamilyStore cfs = getCurrentColumnFamilyStore();
+        StorageAttachedIndexGroup group = StorageAttachedIndexGroup.getIndexGroup(cfs);
+        StorageAttachedIndex index = (StorageAttachedIndex) group.getIndexes().iterator().next();
+        SSTableReader sstable = Iterables.getOnlyElement(cfs.getLiveSSTables());
+
+        Set<Component> components = StorageAttachedIndexGroup.getLiveComponents(sstable, getIndexesFromGroup(group));
+        assertEquals(Version.LATEST.onDiskFormat().perSSTableIndexComponents().size() + 1, components.size());
+
+        // index files are released but not removed
+        cfs.invalidate(true, false);
+        Assert.assertTrue(index.getIndexContext().getView().getIndexes().isEmpty());
+        for (Component component : components)
+            Assert.assertTrue(sstable.descriptor.fileFor(component).exists());
+    }
+
+    @Test
+    public void getLiveComponentsForEmptyIndex() throws Throwable
+    {
+        createTable("CREATE TABLE %s (pk int primary key, value text)");
+        createIndex("CREATE CUSTOM INDEX ON %s(value) USING 'StorageAttachedIndex'");
+        waitForIndexQueryable();
+        execute("INSERT INTO %s (pk) VALUES (1)");
+        flush();
+
+        ColumnFamilyStore cfs = getCurrentColumnFamilyStore();
+        StorageAttachedIndexGroup group = StorageAttachedIndexGroup.getIndexGroup(cfs);
+        Set<SSTableReader> sstables = cfs.getLiveSSTables();
+
+        assertEquals(1, sstables.size());
+
+        Set<Component> components = StorageAttachedIndexGroup.getLiveComponents(sstables.iterator().next(), getIndexesFromGroup(group));
+
+        assertEquals(Version.LATEST.onDiskFormat().perSSTableIndexComponents().size() + 1, components.size());
+    }
+
+    @Test
+    public void getLiveComponentsForPopulatedIndex() throws Throwable
+    {
+        createTable("CREATE TABLE %s (pk int primary key, value text)");
+        IndexContext indexContext = createIndexContext(createIndex("CREATE CUSTOM INDEX ON %s(value) USING 'StorageAttachedIndex'"), UTF8Type.instance);
+        waitForIndexQueryable();
+        execute("INSERT INTO %s (pk, value) VALUES (1, '1')");
+        flush();
+
+        ColumnFamilyStore cfs = getCurrentColumnFamilyStore();
+        StorageAttachedIndexGroup group = StorageAttachedIndexGroup.getIndexGroup(cfs);
+        Set<SSTableReader> sstables = cfs.getLiveSSTables();
+
+        assertEquals(1, sstables.size());
+
+        Set<Component> components = StorageAttachedIndexGroup.getLiveComponents(sstables.iterator().next(), getIndexesFromGroup(group));
```

Comment:
Nit: we can prevent the "might be null" warning with an `Assert.assertNotNull(group)` check.